### DISCUSSION TITLE
Where to store global constants?

### DISCUSSION THREAD

**Discussion by wsw70** [quasarframework/quasar#10090] [General - Components / Directives / etc] [answered]
I am building a home-use, one-off application (Quasar V2) with one-off components (they will not be reused anywhere else). 

All of the components need to access a global constant (that I want to make global to make my development easier, breaking the portability/re-usability of the components) and I was wondering where I could define that global constant (and how to retrieve it within the components)?

My current solution is to `import { apiUrl } from 'assets/config'` where `apiUrl` is defined as `export const apiUrl = "http://localhost:1495"` in `assets/config.js` - this works but I would like to understand how to deal with global constants (even though it is bad design, I am aware of that)

**Comment by justinbeaty** [comment]
@wsw70 I haven't yet tried out Quasar v2, but in v1 I would do this in a boot file. Like this:

```
import Vue from 'vue'

export default (ctx) => {
  Vue.prototype.$config = ctx.app.$config = ctx.store.$config = ctx.router.$config = { 
    apiUrl: 'https://api.example.com'
  }
})
```

That makes it available pretty much anywhere in the Quasar project (with `this.$config.appUrl`). There's probably a similar way with Vue 3.

**Comment by hawkeye64** [comment, marked as answer]
The best way for Vue 3 is to use provide/inject

In a boot file
```js
import { boot } from 'quasar/wrappers'
import axios from 'axios'

export default boot(({ app }) => {
  app.provide('$axios', axios)
})
```
Where you use it:
```js
const { inject } = Vue;
...
setup() {
  const $axios = inject('$axios');   // injecting in a component that wants it
  // $axios.get(...)
}
```

**Reply by justinbeaty** [reply]
I like this much more than how I've been doing it in v1. I'm looking forward to upgrading!

**Comment by exninja** [comment]
You can also put environment specific constants in quasar.conf.js then let quasar compilation use the correct one depending on quasar dev / quasar build.

```
module.exports = configure(function (ctx) {
  return {
    ...
    build: {
      ...
      env: ctx.dev
      ? { // dev
        API_END_POINT: 'http://localhost:1495',
      }
      : { // production
        API_END_POINT: 'https://production.com',
      }
```

Then reference API_END_POINT wherever you need it.

e.g.

```
axios.defaults.baseURL = process.env.API_END_POINT
```

or...

```
setup () {
  const myApiUrl = process.env.API_END_POINT
```
